src/services/resource.js

```diff
@@ -0,0 +1,1297 @@
+const transformRecordByURLToResource = (records, resources) => (
+  Array.isArray(records)
+  && Array.isArray(resources)
+    ? records
+      .map(({ url, ...resource }) => ({
+        ...resource,
+        resourceId: resources.find((r) => r.url === url)?.id,
+      }))
+      .filter((resource) => resource.resourceId !== undefined
+        && resource.resourceId !== null)
+    : []);
+
+// Compare the incomingResources and the currentResources to generate five sets of modifications:
+// new: completely new records to be added.
+// expanded: records that existed already, but now are referenced in additional sourceFields.
+// delta: records that existed already, but now are referenced in additional sourceFields as well
+//     as not referenced in priorly referenced sourceFields.
+// reduced: records that existed already, but now not referenced in priorly referenced sourceFields.
+// removed: records that existed already, that need to be removed
+// Each of these are separate to allow for all of them to be executed as modifications to the
+// database in parallel, as each of them will operate on distinct records with no overlap.
+// Preventing any locking contentions between the five datasets. Once completethe five datasets
+// are grouped into the three operations as follows:
+// create: new
+// update: expanded, delta, reduced
+// destroy: remove
+const filterResourcesForSync = (
+  incomingResources,
+  currentResources,
+  ignoreDestroy,
+) => {
+  if (!Array.isArray(incomingResources)
+    || !Array.isArray(currentResources)) {
+    return {
+      create: [],
+      update: [],
+      destroy: [],
+    };
+  }
+  // pull all of the new and expanded resources in a single pass over the incomingResources.
+  const newExpandedResources = incomingResources
+    .reduce((resources, resource) => {
+      const matchingFromFields = currentResources
+        .filter((cr) => cr.genericId === resource.genericId
+        && cr.resourceId === resource.resourceId);
+      const isCreated = matchingFromFields.length === 0;
+      if (isCreated) {
+        const created = resources.created
+          ?.find((r) => r.genericId === resource.genericId
+          && r.resourceId === resource.resourceId);
+        if (created) {
+          created.sourceFields = [...new Set([...created.sourceFields, ...resource.sourceFields])];
+          return resources;
+        }
+        return {
+          created: [
+            ...resources.created,
+            {
+              genericId: resource.genericId,
+              resourceId: resource.resourceId,
+              sourceFields: resource.sourceFields,
+            },
+          ],
+          expanded: resources.expanded,
+        };
+      }
+
+      const isExpanded = matchingFromFields
+        .filter((mff) => resource.sourceFields
+          .filter((l) => mff.sourceFields.includes(l))
+          .length < resource.sourceFields.length)
+        .length > 0;
```

Comment:
the .length < resources.sourceFields.length perplexed me for a bit, but I think what we're doing here is just ensuring that every item in the resource.sourceFields array is included in the mff.sourceFields array. if that's the case I think we can just use every here to again avoid creating a new array just to check its length:

```suggestion
const isExpanded = matchingFromFields.some((mff) => (
  mff.sourceFields.length > resource.sourceFields.length &&
  resource.sourceFields.every((l) => mff.sourceFields.includes(l))
));
```